### PR TITLE
Fixed bug #172. (Rewrote the WaitForExitAsync function and set EnableRaisingEvents correctly)

### PR DESCRIPTION
**Pull request by Tthecreator**
When converting subtitles, there is a function that should wait until ffmpeg(the process) completes it's work (when it exits). The WaitForExitAsync function however, was broken and always waited out it's full timeout time. The function has been rewritten using a custom task instead of a taskSource and I've tested it to work. 

I'm pretty sure I did everything correctly, but I haven't really done much with C# before so I hope my thread/task code is safe. (It should be).

This is to fix bug #172.

### CODE REVIEW

File: Emby.Server.Implementations/Diagnostics/CommonProcess.cs [hunk 1]
```diff
@@ -121,30 +121,29 @@ public bool WaitForExit(int timeMs)
...
         }
 
-        public Task<bool> WaitForExitAsync(int timeMs)
+        private bool WaitForExitAsyncCompletion(CancellationToken cancellationToken)
         {
-            //if (_process.WaitForExit(100))
-            //{
-            //    return Task.FromResult(true);
-            //}
+            while (!HasExited && !cancellationToken.IsCancellationRequested) { }//wait for exit
```
Comment (cvium) [inline Emby.Server.Implementations/Diagnostics/CommonProcess.cs]: This is just busy waiting. Your CPU is not gonna be happy about it.

Comment (Tthecreator) [reply Emby.Server.Implementations/Diagnostics/CommonProcess.cs]: Yea I see now. I'll guess I'll dive deeper into the code to make a change in HasExited trigger something.

**Comment by Tthecreator** [comment]
No, the process definitely exits properly.

**Comment by Tthecreator** [comment]
Well this should work better and it's not all that different from what was there. The bug was that EnableRaisingEvents would only be enabled if it was already enabled. (It wasn't so the code didn't work).
This solution should be 'the one'

File: Emby.Server.Implementations/Diagnostics/CommonProcess.cs [hunk 2]
```diff
@@ -42,14 +43,16 @@ public CommonProcess(ProcessOptions options)
                 StartInfo = startInfo
             };
 
-            if (options.EnableRaisingEvents)
-            {
-                _process.EnableRaisingEvents = true;
-                _process.Exited += _process_Exited;
-            }
+
+            _process.EnableRaisingEvents = true;
```
Comment (cvium) [inline Emby.Server.Implementations/Diagnostics/CommonProcess.cs]: I don't think this is the right solution. We probably still want it to be configurable.

It should be changed here in options instead: https://github.com/jellyfin/jellyfin/blob/master/MediaBrowser.MediaEncoding/Subtitles/SubtitleEncoder.cs#L449

File: Emby.Server.Implementations/Diagnostics/CommonProcess.cs [hunk 3]
```diff
@@ -123,28 +126,24 @@ public bool WaitForExit(int timeMs)
...
             var cancellationToken = new CancellationTokenSource(timeMs).Token;
 
-            if (HasExited)
-            {
-                return Task.FromResult(true);
-            }
+            _process.Exited += (sender, args) => taskSource.TrySetResult(true);
 
-            _process.Exited += (sender, args) => tcs.TrySetResult(true);
+            cancellationToken.Register(()=>taskSource.TrySetResult(HasExited));
```
Comment (cvium) [inline Emby.Server.Implementations/Diagnostics/CommonProcess.cs]: Missing spaces around `=>`

File: Emby.Server.Implementations/Diagnostics/CommonProcess.cs (hunk 2 above)
Comment (Tthecreator) [reply Emby.Server.Implementations/Diagnostics/CommonProcess.cs]: I see, guess I did an oopsie again. Fixed now.

**Comment by JustAMan** [comment]
@Tthecreator please rebase your PR against `dev` branch, not `master`.

**Comment by Tthecreator** [comment]
> @Tthecreator please rebase your PR against `dev` branch, not `master`.

Looks like I need to make a new pull request since I cant change my source from master to dev.
Sorry for all the inconvenience. The new request is at #671.

**Comment by joshuaboniface** [comment]
@Tthecreator No problem, I'll close this one out!

**Pull request closed without merging**